### PR TITLE
MM-50427: Make MM survive DB replica outage

### PR DESCRIPTION
**Pull request by agnivade**
We monitor the health of DB replicas, and on a fatal error,
take them out of the pool.

On a separate goroutine, we keep pinging the unhealthy replicas,
and on getting a good response back, we add them back to the pool.

https://mattermost.atlassian.net/browse/MM-50427

```release-note
Mattermost is now resilient against DB replica outages and will
dynamically choose a replica if it's alive.

Also added a config parameter ReplicaMonitorIntervalSeconds
whose default value is 5. This controls how frequently unhealthy
replicas will be monitored for liveness check.
```

### CODE REVIEW

File: server/channels/store/sqlstore/store.go
Comment (isacikgoz) [inline server/channels/store/sqlstore/store.go:250]: Is there any point of the attempts here as we returning `dbsql.Open` regardless of what atempt number is.

File: server/channels/store/sqlstore/store.go
Comment (isacikgoz) [inline server/channels/store/sqlstore/store.go:481]: I didn't get this comment, can you elaborate?

File: server/channels/store/sqlstore/store.go
Comment (isacikgoz) [inline server/channels/store/sqlstore/store.go:530]: Would it make sense if we also make this attempt number configurable?

**Review by isacikgoz — COMMENTED** [review]
Did a quick pass, this is a neat feature but I have some points to discuss.

File: server/channels/store/sqlstore/store.go
Comment (agnivade) [reply server/channels/store/sqlstore/store.go:481]: Sorry, this should be "If all search replicas are down, then go with replica"

File: server/channels/store/sqlstore/store.go
Comment (agnivade) [reply server/channels/store/sqlstore/store.go:250]: From the documentation:

> // Open may just validate its arguments without creating a connection
// to the database. To verify that the data source name is valid, call
// Ping.

So basically, `sql.Open` might not actually do anything worth retrying. `Ping` is the main thing.

File: server/channels/store/sqlstore/store.go
Comment (agnivade) [reply server/channels/store/sqlstore/store.go:530]: IMO, I wouldn't want to. This anyways runs in a loop with the interval controlled by the `ReplicaMonitorIntervalSeconds`. If a customer wants fast recovery, they can set it to a low number. Or if they want slow recovery, they can increase it. That should be the setting to tune.

File: server/channels/store/sqlstore/sqlx_wrapper.go
Comment (agnivade) [reply server/channels/store/sqlstore/sqlx_wrapper.go:485]: I wanted to keep a single `return err` statement. Otherwise, technically you can return after any of the `w.isOnline.Store(false)` statements, and then it gets complicated with named returns and such.

File: server/channels/store/sqlstore/sqlx_wrapper.go
Comment (agnivade) [reply server/channels/store/sqlstore/sqlx_wrapper.go:484]: Readability mainly. But 0/5 if you prefer to remove.

File: server/channels/store/sqlstore/sqlx_wrapper.go
Comment (agnivade) [reply server/channels/store/sqlstore/sqlx_wrapper.go:506]: Yes, definitely a good point re: mysql. I was trying to rely on the error message to make this decision. Postgres makes it easy by having a Fatal error category. But not so much with MySQL.

The actual network connectivity issue is already handled with the `if errors.As(err, &netError) && (!netError.Temporary() && !netError.Timeout())` condition. This was me just being more comprehensive.

If you think it's not necessary (and I am starting to agree as well, atleast for MYSQL), we can remove it.

File: server/channels/store/sqlstore/store.go
Comment (agnivade) [reply server/channels/store/sqlstore/store.go:259]: re-arranged

File: server/channels/store/sqlstore/store.go
Comment (agnivade) [reply server/channels/store/sqlstore/store.go:475]: Thanks. Looks much better now.

File: server/channels/store/sqlstore/store.go
Comment (agnivade) [reply server/channels/store/sqlstore/store.go:546]: Config changes to DSNs aren't reloadable at run time. There are no config listeners attached to sqlstore, so I think it should be fine.

**Comment by agnivade** [comment]
@streamer45 @isacikgoz - PTAL.

File: server/channels/store/sqlstore/store.go
Comment (isacikgoz) [reply server/channels/store/sqlstore/store.go:250]: TIL, thanks for clarification.

File: server/channels/store/sqlstore/store.go
Comment (isacikgoz) [reply server/channels/store/sqlstore/store.go:517]: Makes sense, thanks.

File: server/channels/store/sqlstore/sqlx_wrapper.go
Comment (streamer45) [reply server/channels/store/sqlstore/sqlx_wrapper.go:506]: Personally I am more worried about potentially marking something as offline when in fact it's not. But again, I am not sure how aggressive we want to be with this, is it better to be wrong in one direction (incorrectly set it offline) or the other (fail to detect it's offline)? No strong opinions here, will leave it to you.

**Review by isacikgoz — APPROVED** [review]
This looks good, very nice implementation. Would it make sense to run a load test before merging? Maybe we can get some idea to address @streamer45's concern on being aggressive. We can observe the false positives in a congested server.

File: server/channels/store/sqlstore/sqlx_wrapper.go
Comment (agnivade) [reply server/channels/store/sqlstore/sqlx_wrapper.go:506]: I am leaning towards not being so aggresive here. Will update.

**Comment by agnivade** [comment]
Now you have to approve again @isacikgoz :P

**Review by isacikgoz — DISMISSED** [review]
> Now you have to approve again @isacikgoz :P

Haha it was a one time thing /s :p

File: server/channels/store/sqlstore/store.go
Comment (streamer45) [inline server/channels/store/sqlstore/store.go:53]: I see that this is used in two slightly different cases. First is the actual context timeout to ping the db and the other one is more like an interval to wait before attempting to ping again. I think this is totally fine for the latter use but I wonder if for the former one could be a bit too low to potentially be causing unwanted failures (e.g. DNS resolution can take a while).

Comment (agnivade) [reply server/channels/store/sqlstore/store.go:53]: Ahh sorry my bad, that was a debugging change I meant to revert. Only the pingAttempts was meant to be reduced to 5. 18 seemed like a weird number. Nice catch!

**Review by streamer45 — APPROVED** [review]
Looks great, thanks :+1:

**Comment by agnivade** [comment]
Except .. now we need @isacikgoz to approve for the 3rd time.

**Comment by streamer45** [comment]
> Except .. now we need @isacikgoz to approve for the 3rd time.

Fun times :)

**Review by isacikgoz — APPROVED** [review]
I hereby approve this PR with my full consent. (I hope it will be enough this time)

**Comment by agnivade** [comment]
>  Would it make sense to run a load test before merging? Maybe we can get some idea to address @streamer45's concern on being aggressive.

I tuned it down to be non-aggressive. Now only permanent network errors will trigger a removal of a replica.

**Pull request merged**
